File: schema_salad/java/main_utils/UnionLoader.java
package ${package}.utils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class UnionLoader implements Loader<Object> {
  private final List<Loader> alternates;

  public UnionLoader(List<Loader> alternates) {
    this.alternates = alternates;
  }

  public UnionLoader(Loader[] alternates) {
    this(Arrays.asList(alternates));
  }

  public Object load(
      final Object doc,
      final String baseUri,
      final LoadingOptions loadingOptions,
      final String docRoot) {
    final List<ValidationException> errors = new ArrayList();
    for (final Loader loader : this.alternates) {
      try {
        return loader.load(doc, baseUri, loadingOptions, docRoot);
      } catch (ValidationException e) {
        errors.add(e);
      }
    }
    throw new ValidationException("Failed to match union type", errors);
  }
}

File: schema_salad/typescript/util/loaders/unionLoader.ts
import { Loader, LoadingOptions, Saveable, ValidationException } from '../internal'

export class UnionLoader implements Loader {
  alternates: Loader[]

  constructor (alternates: Loader[]) {
    this.alternates = alternates
  }

  async load (doc: any, baseuri: string, loadingOptions: LoadingOptions, docRoot?: string): Promise<Saveable> {
    const errors: ValidationException[] = []
    for (const t of this.alternates) {
      try {
        return await t.load(doc, baseuri, loadingOptions)
      } catch (e) {
        if (e instanceof ValidationException) {
          errors.push(new ValidationException(`tried ${t.constructor.name} but`, [e]))
        }
      }
    }
    throw new ValidationException('', errors).withBullet('-')
  }
}
